[docs] Add a guide for setting dark mode by default

siriwatknp opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

**Preview**: https://deploy-preview-34839--material-ui.netlify.app/joy-ui/customization/dark-mode/

From the [docs feedback](https://mui-org.slack.com/archives/C041SDSF32L/p1666234661983459).

> How do I set global dark mode by default?

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

danilo-leal commented on docs/data/joy/guides/applying-dark-mode/applying-dark-mode.md:
```diff
@@ -105,3 +105,17 @@ export function onRenderBody({ setPreBodyComponents }) {
+
+## Dark mode by default
+
+Set `defaultMode: 'dark'` to use dark color scheme for the first time users.
```
```suggestion
To have dark mode as the default theme of your app, set `defaultMode: 'dark'`. 
```

Small copy iteration.

oliviertassinari reviewed (commented):
👍

Thoughts:

1. A demo could help a lot to be able to build reproductions that are dark mode only. 
2. I wonder if this isn't more important than the rest of the page, so could be closer to the top.
3. The dark mode docs page is under "How to guide" but all the pages that aren't API references are also how to guides in our docs, only that they are groups in "logical" sections. So per this rational, I think that it could be better to have the docs about the dark mode under the Customization section.

siriwatknp commented:
> 👍
> 
> Thoughts:
> 
> 1. A demo could help a lot to be able to build reproductions that are dark mode only. It could also serve as a visual regression test case.
> 2. I wonder if setting the default is more important than the rest of the page, so maybe it should be closer to the top.
> 3. The dark mode docs page is under "How to guide" but all the pages that aren't API references are also how to guides in our docs, only that they are groups in "logical" sections. So per this rational, I think that it could be better to have the docs about the dark mode under the Customization section. Meaning, we could use "How to guide" for pages we don't know where else to put.

Agree on all points, let me update them altogether.

oliviertassinari commented on docs/data/joy/customization/dark-mode/DarkModeByDefault.js:
```diff
@@ -0,0 +1,41 @@
+import Sheet from '@mui/joy/Sheet';
+import Chip from '@mui/joy/Chip';
+import Typography from '@mui/joy/Typography';
+
+const useEnhancedEffect =
+  typeof window !== 'undefined' ? React.useLayoutEffect : React.useEffect;
```
Would it be simpler?

```suggestion
import { unstable_useEnhancedEffect as useEnhancedEffect } from '@mui/utils';

```

I assume that we will want to make this module stable, since we use it everywhere.

siriwatknp replied:
> Would it be simpler?

I don't think so. The first time I see this import I have no idea what it does so I have to look into the source code anyway. It is even worse for developers who want to copy the code or open the CodeSandbox.

oliviertassinari replied:
Oh, interesting, agree that there is value in inlining to be clearer about what happens.

oliviertassinari commented on docs/data/joy/customization/dark-mode/DarkModeByDefault.js:
```diff
@@ -0,0 +1,41 @@
+  typeof window !== 'undefined' ? React.useLayoutEffect : React.useEffect;
+
+export default function ButtonThemes() {
+  // the `node` is used for attaching CSS variables to this demo, you might not need it in your app.
```
It overflows in the demo:

```suggestion
  // the `node` is used for attaching CSS variables to this demo, you might not
  // need it in your app.
```

<img width="795" alt="Screenshot 2022-10-24 at 12 50 11" src="https://user-images.githubusercontent.com/3165635/197509734-7a12abf4-d070-4f56-9315-8ee444a94d3e.png">

oliviertassinari commented on docs/data/joy/customization/dark-mode/DarkModeByDefault.js:
```diff
@@ -0,0 +1,41 @@
+export default function ButtonThemes() {
+  // the `node` is used for attaching CSS variables to this demo, you might not need it in your app.
+  const [node, setNode] = React.useState(null);
+  useEnhancedEffect(() => {
+    setNode(document.getElementById('dark-mode-by-default'));
+  }, []);
```
**Off topic**

This could suggest that nesting dark with a light theme in Joy UI, might be a bit of a struggle (it's not a single rendering step).

oliviertassinari commented on docs/data/joy/customization/dark-mode/DarkModeByDefault.js:
```diff
@@ -0,0 +1,41 @@
+const useEnhancedEffect =
+  typeof window !== 'undefined' ? React.useLayoutEffect : React.useEffect;
+
+export default function ButtonThemes() {
```
```suggestion
export default function DarkModeByDefault() {
```

oliviertassinari reviewed (commented):
Would it make sense to have this demo and the others in TypeScript? It could serve as a good stress test to spot TypeScript issues.

siriwatknp commented:
@oliviertassinari it is ready for a second review.

- added typescript
- added "[identify the system mode](https://deploy-preview-34839--material-ui.netlify.app/joy-ui/customization/dark-mode/#identify-the-system-mode)" demo.

danilo-leal reviewed (approved):
This is looking great, thanks for putting it up, Jun! Sam will definitely do a thorough copy review but I already took the opportunity to do some content rearranging + writing tweaks. I wonder how to make it better when similar and close sections have exact same content, though.

oliviertassinari commented on docs/data/joy/customization/dark-mode/dark-mode.md:
````diff
@@ -1,8 +1,58 @@
+getInitColorSchemeScript({ defaultMode: 'dark' });
+```
+
+## Matching device's mode preference
````
Maybe to make the headers easier to skip read.

```suggestion
## Matching device's preference
```

oliviertassinari commented on docs/data/joy/customization/dark-mode/dark-mode.md:
```diff
@@ -1,8 +1,58 @@
+
+## Matching device's mode preference
+
+To have your app matching the prefferred mode the user has set on their devide, use `defaultMode: 'system'`.
```
A couple of typos:

```suggestion
Use `defaultMode: 'system'` to have your app match the preferred mode the user has set on their device.
```

samuelsycamore replied:
Iterating on the suggestion above:

```suggestion
Use `defaultMode: 'system'` to set your app's default mode to match the user's chosen preference on their device.
```

oliviertassinari commented on docs/data/joy/customization/dark-mode/dark-mode.md:
```diff
@@ -1,8 +1,58 @@
+# Dark mode
 
-<p class="description">A how-to-guide for applying dark mode to your application with Joy UI.</p>
+<p class="description">Learn about the different methods for having dark mode with Joy UI.</p>
```
Not sure:
```suggestion
<p class="description">Learn about the different methods for enabling dark mode with Joy UI.</p>
```

oliviertassinari commented on docs/data/joy/customization/dark-mode/dark-mode.md:
```diff
@@ -1,8 +1,58 @@
+
+### Identify the system mode
+
+To check if the user's preference is in light or dark mode, use the `useColorScheme` React hook:
```
To be more direct?

```suggestion
Use the `useColorScheme` React hook to check if the user's preference is in light or dark mode:
```

oliviertassinari reviewed (commented):
It looks good 👍

samuelsycamore commented on docs/data/joy/customization/dark-mode/dark-mode.md:
```diff
@@ -1,8 +1,58 @@
+
+{{"demo": "DarkModeByDefault.js"}}
+
+For server-side applications, check out the framework setup in the [guide](#server-side-rendering) below and use the same value to the `getInitColorSchemeScript` function.
```
```suggestion
For server-side applications, check out the framework setup in [the section below](#server-side-rendering) and provide the same value to the `getInitColorSchemeScript` function:
```

samuelsycamore commented on docs/data/joy/customization/dark-mode/dark-mode.md:
```diff
@@ -1,8 +1,58 @@
+{{"demo": "IdentifySystemMode.js"}}
+
+:::warning
+**Note:** Make sure to use `useColorScheme()` in a component that's inside `<CssVarsProvider>`, otherwise it will throw an error.
```
```suggestion
The `useColorScheme()` hook only works with components nested inside of `<CssVarsProvider>`—otherwise it will throw an error.
```

samuelsycamore commented on docs/data/joy/customization/dark-mode/dark-mode.md:
```diff
@@ -30,7 +80,7 @@ function ModeToggle() {
 **Note:** Make sure to use `useColorScheme()` in a component that's inside `<CssVarsProvider>`, otherwise it will throw an error.
```
```suggestion
The `useColorScheme()` hook only works with components nested inside of `<CssVarsProvider>`—otherwise it will throw an error.
```

mnajdova commented on docs/data/joy/customization/dark-mode/DarkModeByDefault.js:
```diff
@@ -0,0 +1,33 @@
+  return (
+    <CssVarsProvider
+      defaultMode="dark"
+      // these props are specific to this demo, you might not need them in your app.
```
A short descriptive of the props could help a lot, so that people don't need to switch to the `CssVarsProvider`'s API to check when they might need them

siriwatknp replied:
Added!

siriwatknp commented:
@samuelsycamore @danilo-leal Wow 😲, it looks much much better. The content flow is so smooth.

oliviertassinari commented on docs/data/joy/customization/dark-mode/ModeToggle.tsx:
```diff
@@ -0,0 +1,60 @@
+}
+
+export default function ModeToggle() {
+  // the `node` is used for attaching CSS variables to this demo, you might not need it in your application.
```
Scroll overflow, I would recommend shortening:

<img width="812" alt="Screenshot 2022-11-07 at 02 15 20" src="https://user-images.githubusercontent.com/3165635/200207271-1953e8a9-e0d0-4701-bd88-9ca36ed77037.png">

```suggestion
  // the `node` is used for attaching CSS variables to this demo, you might not need
  // it in your application.
```

*x the other comments*

oliviertassinari commented on docs/data/joy/customization/dark-mode/ModeToggle.tsx:
```diff
@@ -0,0 +1,60 @@
+          flexGrow: 1,
+          p: 2,
+          m: -3,
+          borderRadius: 'sm',
```
To remove the border on mobile?

<img width="414" alt="Screenshot 2022-11-07 at 02 17 06" src="https://user-images.githubusercontent.com/3165635/200207386-9fc5afd4-dc2f-4765-a585-7cbc0f699194.png">

```suggestion
          borderRadius: [0, 'sm'],
```

siriwatknp replied:
Nice catch.

oliviertassinari commented on docs/data/joy/customization/dark-mode/DarkModeByDefault.tsx:
```diff
@@ -0,0 +1,33 @@
+    <CssVarsProvider
+      defaultMode="dark"
+      // the props below are specific to this demo, you might not need them in your app.
+      colorSchemeSelector="#demo_dark-mode-by-default" // the selector to apply CSS theme variables stylesheet.
```
If we care about the debugging use case here, I think that it would be better to have an intermediary `<div>` to host the styles. When you open the dev tools to find what are the styles of `<Sheet>`, we get flooded with CSS variables properties:

<img width="332" alt="Screenshot 2022-11-07 at 02 19 56" src="https://user-images.githubusercontent.com/3165635/200207613-05d8a3aa-931b-42f9-8bc2-92666b70e86e.png">

siriwatknp replied:
Done!
